.github/workflows/build.yml

```
@@ -44,14 +46,27 @@ jobs:
       uses: actions/upload-artifact@v2
       with:
         name: python-packages
-        path: |
-          dist/*.whl
+        path: dist/*.whl
     - name: Install wheel
       run: pip install dist/*.whl
-    - run: pip install -r requirements-dev.txt
-    - name: Build docs
-      run: sphinx-build . _build
+    - name: Build html docs
       working-directory: docs
+      run: sphinx-build . _build/html
+    - name: Build latex docs
+      working-directory: docs
+      run: sphinx-build -b latex . _build/latex
+    - name: Upload doc builds as artifacts
+      uses: actions/upload-artifact@v2
+      with:
+        name: doc-builds
+        path: docs/_build/
+    - name: upload docs to github pages
+      # if: ${{ github.event_name == 'release'}}
+      if: ${{ github.event_name != 'pull_request'}}
```

Should only upload for main branch.

Also should use explicit hash for all third-party github actions.

---

> Also should use explicit hash for all third-party github actions.

I was wondering why you do that. Using `@v3` helps the author keep patches being distributed without users needing to change their tag.

---

Kinda like this? I think the condition below (in the python-publish-package job) will only work if the branch is named master (not main).
```suggestion
      if: ${{ (github.event_name != 'pull_request') && (github.ref == 'refs/heads/main') }}
```

---

Yeah, should be changed to be `github.event_name == 'push' && github.ref == 'refs/heads/main'`

The condition in `python-publish-package` should also be updated.

---

I changed the condition in `python-publish-package` job and my fork _tried_ to publish to pypi (the `||` bypassed the check for a tagged commit). I have enabled actions on my fork for linting reasons.

Is it ok if we add `release` event triggers and then only publish to pypi on release events? Alternatively I think this might be solved by comparing `refs/heads/jbms/main` instead(?) or making sure that you triggered the workflow with the env var `GITHUB_ACTOR`.

---

>being pushed to the Test PyPI server --- users can install from there if they want the latest development version

Ah. That explains why you don't use `twine check`. I should've been trying this for my tests on my RTD project.

> Not sure what the tradeoffs are

I've often seen a sorely lacking release description (if any) when people manually push a tag. I often look to the release description for a list of changes, but maybe that could be supplemented with a automated change-log action (there's a few out there) - probably better addressed at a later time.

If the intention is to publish dev versions, then why limit publishing (for the test pypi server) to just the main branch? I can understand this intention for the regular pypi server.

---

Branches other than main may just be work in progress, may be force pushed, etc.  Additionally, the versions published to the test pypi server will have a version number derived similarly to `git describe`, based on:
- the latest tag that contains the commit
- the number of commits since then

For a single branch without force pushes that ensures a unique, monotonically increasing version number.  But if you have multiple branches then the version numbers won't be unique.

---

That actually better explains the comments in 
https://github.com/jbms/sphinx-immaterial/blob/d59f75c815f7018e2a651c8b0c3741db42304c9c/setup.py#L195-L202

ok, so I'll change the condition to something like
```yml
  python-publish-package:
    # Only publish package on push to tag or default branch of jbms/sphinx-immaterial.
    if: ${{ github.event_name == 'push' && github.repository == 'jbms/sphinx-immaterial' && (startsWith(github.ref, 'refs/tags/v') || github.ref == 'refs/heads/main') }}
```